main/resource-access.bs

```diff
@@ -130,31 +125,150 @@ Note:
+
+Servers MUST support the HTTP `GET`, `HEAD` and `OPTIONS` methods [[!RFC7231]]
+for clients to read resources or to determine communication options.
+[[Source](https://github.com/solid/specification/issues/39#issuecomment-538017667)]
+
+Servers MUST indicate their support for HTTP Methods by responding to HTTP
+`GET`, `HEAD` and `OPTIONS` requests on the resource's URL with the HTTP
+Method tokens in the HTTP response header `Allow`. Servers MUST indicate
+supported media types in the `Accept-Patch` [[!RFC5789]] and the `Accept-Post`
+[[!LDP]] headers for matching HTTP methods in `Allow` header value.
+
+Servers MAY include the HTTP `Accept-Patch` and `Accept-Post` headers in the
+response of a `OPTIONS *` request.
 
-Servers MUST allow creating new resources with a POST request to URI path
+[[Source](https://github.com/solid/specification/issues/85#issuecomment-575386251)]
+[[Source](https://github.com/solid/specification/issues/43)]
+
+### Writing resources ### {#write}
+
+When a server supports the HTTP `PUT`, `POST` and `PATCH` methods [[!RFC7231]]
+this specification imposes the following requirements:
+[[Source](https://github.com/solid/specification/issues/39#issuecomment-538017667)]
+
+Servers MUST create intermediate containers and include corresponding
+containment triples in container representations derived from the URI path
+component of `PUT`, `POST` and `PATCH` requests.
+[[Source](https://github.com/solid/specification/issues/68#issuecomment-561690124)]
+
+Servers MUST allow creating new resources with a `POST` request to URI path
 ending `/`. Servers MUST create a resource with URI path ending `/{id}` in
 container `/`. Servers MUST create a container with URI path ending `/{id}/`
 in container `/` for requests including the HTTP `Link` header with
 `rel="type"` targeting a valid LDP container type. Servers MUST handle
 subsequent requests to the newly created container's URI as if it is a valid
-LDP container type by including it the HTTP response's `Link` header.
+LDP container type by including the HTTP response's `Link` header.
+[[Source](https://github.com/solid/specification/pull/160#issuecomment-636822687)]
+
+When a `POST` method request targets a non-container resource without an
+existing representation, the server `MUST` respond with the `404` status code.
+[[Source](https://github.com/solid/specification/issues/108#issuecomment-549448159)]
+
+When a `PUT` or `PATCH` method request targets an auxiliary resource, the
+server MUST create or update it. When a `POST` method request with the `Slug`
+header targets an auxiliary resource, the server MUST respond with the `403`
+status code and response body describing the error.
+[[Source](https://github.com/solid/specification/issues/42#issuecomment-616688848)]
+
+Servers SHOULD NOT allow HTTP `POST`, `PUT` and `PATCH` to update a
+container's containment triples; if the server receives such a request, it
+SHOULD respond with a `409` status code.
```

Comment:
Is it legit for a server to simply ignore any user-supplied containment triples? For instance, say you have some user-managed triples in a container (along with some containment triples). Assume that you have two clients manipulating the data of that container. The first client fetches the RDF representation of that container (which includes the containment triples and an `<> rdfs:label "some label"` triple). The second client `POST`s to the container, adding a new child resource. Then the first client issues a `PUT` (missing the newly created child resource but with the updated `rdfs:label` value). Assume the clients are not using `If-Match` headers.

Should the sever accept the first client's PUT request (and just ignore the containment triples)? Or is it the responsibility of the client to remove those triples from the request? Or must a client use PATCH in this case? Or must a client use `If-Match` headers in all cases? (the section below suggests that conditional requests are at the `MAY` level)

Reply:
If we are still in agreement on https://github.com/solid/specification/issues/40#issuecomment-573358652 , I think it entails that 409 will precede conditional errors (412, and probably for 428) as per https://tools.ietf.org/html/rfc7232#section-5 :

>   redirects and failures take precedence over the evaluation of
>   preconditions in conditional requests.

Does that make sense? If so, to take it a step further:

it may also make sense that the client should follow up with `If-Match` on their next write request. Of course they'll need to get a hold of the latest resource state before doing so. Without additional instructions to give to a client, it is possible that a simple client will not get its update through when the validator on a resource changes since their last interaction with the resource. Server could inform exactly what happened and what the client should do next in the response body of a 409 but it is not entirely reliable if this were to be automated - we'd have to bake in additional know-how on how clients can make use of structured problem details.

I'm also not confident about how sensible it'd be for a server to use an ETag in the response of a 409 for a write request.

Alternatively, we can revisit https://github.com/solid/specification/pull/188/commits/49cd2f18e55f21df339226cf8e63c5f5424354e5 in that client should use `If-Match` in requests if server includes an ETag with strong validator. However! I'm not aware of if/how clients can make decisions based on strong or weak validator in ETag. So, this may not actually work in practice?

--- 

We arrived here because we didn't indicate weak or strong validator use for servers, and especially on handling RDF sources. Reverting the commit (linked above) was the simple way out.

Reply:
I think we are in agreement w/r/t conditional requests and all the complexities of ETags. What I was specifically interested in with my question relates to triples that are server-managed (e.g. `<> ldp:contains <./resource>`), and whether a server can ignore such server-managed triples on inbound client write requests. A client shouldn't be able to set these triples anyway and so (presumably) a server would need some mechanism to distinguish between those triples and other, client-managed triples.

To be clear, assume you have a resource called `./container/`. A GET requests produces the following:

```turtle
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix ldp:   <http://www.w3.org/ns/ldp#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

<>  rdf:type     ldp:BasicContainer ;
    rdfs:label "A Conttainer"@en ;
    ldp:contains <./resource> .
```

But I realize that the label was misspelled, so I want to perform a PUT to replace that label with the following Turtle: 

```turtle
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix ldp:   <http://www.w3.org/ns/ldp#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

<>  rdf:type     ldp:BasicContainer ;
    rdfs:label "A Container"@en ;
    ldp:contains <./resource> .
```

Assume that the `rdf:type` and `ldp:contains` triples are managed by the server. And assume that the client request isn't attempting to modify any containment triples.

But imagine that, between the GET and the PUT request, another agent has created a second child resource by POSTing to that container. Now there are two `ldp:contains` triples. That means that the PUT request by our client is now out of sync with the state of the container (a conditional request `If-Match` would catch that, but for the sake of argument, we're not using conditional requests). Specifically, the containment triples in the client request are not accurate from the perspective of the server, it is ambiguous whether a client is actually attempting to "update a container's containment triples".

The `SHOULD` language is probably sufficient, but I'm trying to tease out whether, for server-managed triples (such as containment) a server implementation can simply ignore those client-supplied triples or whether the server needs to respond with a 409 Conflict error.

Reply:
This is great feedback, thanks!

Servers should not allow clients to modify containment triples. Put another way, two subsequent container states can't have different containment triples by way of client attempting to update the container. Server responds with 409.

The spec essentially states that clients MAY use `If-Match` by way of implementing RFC 7232's client parts. In order to encourage clients to opt-in to using the `If-Match` header, I've added an optional for servers to use the strong validator on RDF bearing representations: https://github.com/solid/specification/pull/188/commits/b519bf83c0962dfdd17647f946315e1cacaeac3c

Regarding the (SHOULD NOT) requirement level: Currently there is a requirement to have all resources be discoverable and no orphaned resources can be created ( https://github.com/solid/specification/issues/97#issuecomment-547459396 ). It follows that while SHOULD NOT may be sufficient, MUST NOT can ensure that all resources are discoverable. SHOULD NOT permits the system to have "hidden" resources by way of client removing a containment statement while leaving the resource (it contained just earlier) untouched. So, is there a reason to use SHOULD NOT?

Reply:
I agree that MUST NOT would ensure that all resources are discoverable. I would favor that approach.
In other words, I don't have a use case that would suggest using SHOULD NOT, though I'm also not against the use of SHOULD NOT.

Reply:
Let's go with MUST NOT at least towards spec Draft until a case for SHOULD NOT is clear: https://github.com/solid/specification/pull/188/commits/d7b8039691f192fa58c53783f3bd4f13c8bde50e . Thanks.